File: android/Streamo/app/src/main/java/com/streamo/app/tmdb/TmdbCacheKey.kt
package com.streamo.app.tmdb

/**
 * Builder di chiavi deterministiche per la cache TMDB. Ogni chiave include
 * tutti i parametri che cambiano la risposta (endpoint+page, type+id,
 * tvId+season, query+page+genres, sortBy) così chiamate identiche condividono
 * la riga mentre un qualunque param diverso fa miss. I prefissi coincidono con
 * i `TYPE_*` di [TmdbCacheTtl] per consentire l'eviction L1 per categoria.
 */
object TmdbCacheKey {
    fun list(endpoint: String, page: Int) = "${TmdbCacheTtl.TYPE_LIST}:$endpoint:p$page"
    fun details(type: String, id: Int) = "${TmdbCacheTtl.TYPE_DETAILS}:$type:$id"
    fun season(tvId: Int, season: Int) = "${TmdbCacheTtl.TYPE_SEASON}:$tvId:$season"
    fun recommendations(type: String, id: Int) = "${TmdbCacheTtl.TYPE_RECOMMENDATIONS}:$type:$id"
    fun reviews(type: String, id: Int) = "${TmdbCacheTtl.TYPE_REVIEWS}:$type:$id"
    fun searchMulti(query: String, page: Int) = "${TmdbCacheTtl.TYPE_SEARCH}:multi:${query.lowercase()}:p$page"
    fun searchMovie(query: String, page: Int, genres: String?) =
        "${TmdbCacheTtl.TYPE_SEARCH}:movie:${query.lowercase()}:p$page:g${genres ?: ""}"
    fun searchTv(query: String, page: Int, genres: String?) =
        "${TmdbCacheTtl.TYPE_SEARCH}:tv:${query.lowercase()}:p$page:g${genres ?: ""}"
    fun discover(media: String, page: Int, genres: String?, sortBy: String) =
        "${TmdbCacheTtl.TYPE_DISCOVER}:$media:p$page:g${genres ?: ""}:s$sortBy"
    const val genres = "${TmdbCacheTtl.TYPE_GENRES}:merged"

    /** True se la chiave L1 appartiene alla categoria `type` (prefisso). */
    fun matchesType(key: String, type: String): Boolean = key.startsWith("$type:")
}

File: vidaaos/src/data/tmdb/TmdbCacheKey.ts
// Deterministic cache key builders for the TMDB cache. Each key includes
// every parameter that changes the response (endpoint+page, type+id,
// tvId+season, query+page+genres, sortBy) so identical calls share a row
// while any differing param misses. Prefixes match the TYPE_* of
// TmdbCacheTtl to allow L1 eviction by category.
// Verbatim copy of TmdbCacheKey.kt.

import { TmdbCacheTtl } from './TmdbCacheTtl';

export const TmdbCacheKey = {
  list(endpoint: string, page: number): string {
    return `${TmdbCacheTtl.TYPE_LIST}:${endpoint}:p${page}`;
  },
  details(type: string, id: number): string {
    return `${TmdbCacheTtl.TYPE_DETAILS}:${type}:${id}`;
  },
  season(tvId: number, season: number): string {
    return `${TmdbCacheTtl.TYPE_SEASON}:${tvId}:${season}`;
  },
  recommendations(type: string, id: number): string {
    return `${TmdbCacheTtl.TYPE_RECOMMENDATIONS}:${type}:${id}`;
  },
  reviews(type: string, id: number): string {
    return `${TmdbCacheTtl.TYPE_REVIEWS}:${type}:${id}`;
  },
  searchMulti(query: string, page: number): string {
    return `${TmdbCacheTtl.TYPE_SEARCH}:multi:${query.toLowerCase()}:p${page}`;
  },
  searchMovie(query: string, page: number, genres: string | null): string {
    return `${TmdbCacheTtl.TYPE_SEARCH}:movie:${query.toLowerCase()}:p${page}:g${genres ?? ''}`;
  },
  searchTv(query: string, page: number, genres: string | null): string {
    return `${TmdbCacheTtl.TYPE_SEARCH}:tv:${query.toLowerCase()}:p${page}:g${genres ?? ''}`;
  },
  discover(media: string, page: number, genres: string | null, sortBy: string): string {
    return `${TmdbCacheTtl.TYPE_DISCOVER}:${media}:p${page}:g${genres ?? ''}:s${sortBy}`;
  },
  genres: `${TmdbCacheTtl.TYPE_GENRES}:merged`,

  /** True if the L1 key belongs to category `type` (prefix match). */
  matchesType(key: string, type: string): boolean {
    return key.startsWith(`${type}:`);
  },
};

// ponytail: one self-check, fails if key format drifts from Kotlin.
export function demo(): void {
  const k = TmdbCacheKey.list('trending/movie/day', 1);
  // Kotlin: "${TYPE_LIST}:$endpoint:p$page"  -> "list:trending/movie/day:p1"
  console.assert(k === 'list:trending/movie/day:p1', `list key mismatch: ${k}`);
  const d = TmdbCacheKey.details('movie', 42);
  console.assert(d === 'details:movie:42', `details key mismatch: ${d}`);
  const s = TmdbCacheKey.searchMovie('Foo Bar', 2, '28,12');
  console.assert(
    s === 'search:movie:foo bar:p2:g28,12',
    `searchMovie key mismatch: ${s}`,
  );
  const g = TmdbCacheKey.discover('tv', 3, null, 'popularity.desc');
  console.assert(
    g === 'discover:tv:p3:g:spopularity.desc',
    `discover key mismatch: ${g}`,
  );
  console.assert(TmdbCacheKey.matchesType('list:x:p1', 'list') === true);
  console.assert(TmdbCacheKey.matchesType('details:movie:1', 'list') === false);
}